File: desktop/crypto/secure_encryption.py
"""
CryptoJS-Compatible AES Encryption for Python
This module provides the exact same encryption format as CryptoJS AES.encrypt()
"""

import os
import base64
import hashlib
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.primitives import padding
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from cryptography.hazmat.primitives import hashes


class CryptoJSAES:
    """Python implementation that produces identical output to CryptoJS AES.encrypt()"""

    @staticmethod
    def derive_key_and_iv(password: bytes, salt: bytes, key_len: int = 32, iv_len: int = 16):
        """
        Derive key and IV using the same method as CryptoJS (EVP_BytesToKey equivalent)
        This matches CryptoJS.kdf.OpenSSL.execute()
        """
        derived = b''
        h = b''

        while len(derived) < (key_len + iv_len):
            h = hashlib.md5(h + password + salt).digest()
            derived += h

        return derived[:key_len], derived[key_len:key_len + iv_len]

    @staticmethod
    def encrypt(plaintext: str, password: str) -> str:
        """
        Encrypt plaintext with password using CryptoJS-compatible format
        Returns base64 string in CryptoJS format: "Salted__" + salt + encrypted_data
        """
        # Generate random 8-byte salt (CryptoJS standard)
        salt = os.urandom(8)

        # Derive key and IV using CryptoJS method
        key, iv = CryptoJSAES.derive_key_and_iv(password.encode('utf-8'), salt)

        # Pad plaintext using PKCS7
        padder = padding.PKCS7(128).padder()
        padded_data = padder.update(plaintext.encode('utf-8'))
        padded_data += padder.finalize()

        # Encrypt using AES-256-CBC
        cipher = Cipher(algorithms.AES(key), modes.CBC(iv))
        encryptor = cipher.encryptor()
        encrypted = encryptor.update(padded_data) + encryptor.finalize()

        # Format as CryptoJS: "Salted__" + salt + encrypted_data
        salted_prefix = b"Salted__"
        result = salted_prefix + salt + encrypted

        # Return base64 encoded
        return base64.b64encode(result).decode('ascii')

    @staticmethod
    def decrypt(encrypted_data: str, password: str) -> str:
        """
        Decrypt CryptoJS-format encrypted data
        """
        try:
            # Decode base64
            data = base64.b64decode(encrypted_data)

            # Check for "Salted__" prefix
            if not data.startswith(b"Salted__"):
                raise ValueError("Invalid CryptoJS format")

            # Extract salt and encrypted data
            salt = data[8:16]  # 8 bytes after "Salted__"
            encrypted = data[16:]

            # Derive key and IV
            key, iv = CryptoJSAES.derive_key_and_iv(password.encode('utf-8'), salt)

            # Decrypt
            cipher = Cipher(algorithms.AES(key), modes.CBC(iv))
            decryptor = cipher.decryptor()
            padded_plaintext = decryptor.update(encrypted) + decryptor.finalize()

            # Remove padding
            unpadder = padding.PKCS7(128).unpadder()
            plaintext = unpadder.update(padded_plaintext)
            plaintext += unpadder.finalize()

            return plaintext.decode('utf-8')

        except Exception:
            return None


class SecureMnemonicEncryption:
    """Updated class using CryptoJS-compatible encryption"""

    # Mobile-optimized iterations
    PBKDF2_ITERATIONS = 10000
    SALT_SIZE = 16

    def __init__(self):
        pass

    def encrypt_mnemonic(self, mnemonic: str, password: str) -> str:
        """
        Encrypt mnemonic using CryptoJS-compatible format
        """
        if not mnemonic or not mnemonic.strip():
            raise ValueError("Mnemonic cannot be empty")
        if not password or len(password) < 8:
            raise ValueError("Password must be at least 8 characters")

        # Generate salt for PBKDF2
        salt = os.urandom(self.SALT_SIZE)

        # Derive key using PBKDF2 (same as Android)
        kdf = PBKDF2HMAC(
            algorithm=hashes.SHA256(),
            length=32,
            salt=salt,
            iterations=self.PBKDF2_ITERATIONS,
        )
        derived_key = kdf.derive(password.encode('utf-8'))

        # Convert key to hex string (same as Android key.toString())
        key_string = derived_key.hex()

        # Encrypt using CryptoJS-compatible method
        encrypted = CryptoJSAES.encrypt(mnemonic, key_string)

        # Format like Android: salt_hex + ':' + encrypted_data
        salt_hex = salt.hex()
        combined_format = f"{salt_hex}:{encrypted}"

        # Return base64 encoded result
        return base64.b64encode(combined_format.encode('utf-8')).decode('ascii')

    def decrypt_mnemonic(self, encrypted_data: str, password: str) -> str:
        """
        Decrypt mnemonic from Android/CryptoJS format
        """
        try:
            if not encrypted_data or not password:
                return None

            # Decode outer base64
            combined_format = base64.b64decode(encrypted_data).decode('utf-8')

            # Split salt and encrypted data
            parts = combined_format.split(':')
            if len(parts) != 2:
                return None

            salt_hex, encrypted_part = parts

            # Convert salt back to bytes
            salt = bytes.fromhex(salt_hex)

            # Derive the same key
            kdf = PBKDF2HMAC(
                algorithm=hashes.SHA256(),
                length=32,
                salt=salt,
                iterations=self.PBKDF2_ITERATIONS,
            )
            derived_key = kdf.derive(password.encode('utf-8'))

            # Convert key to hex string
            key_string = derived_key.hex()

            # Decrypt using CryptoJS-compatible method
            decrypted = CryptoJSAES.decrypt(encrypted_part, key_string)

            return decrypted

        except Exception:
            return None

    def verify_mnemonic_format(self, mnemonic: str) -> bool:
        """Validate mnemonic format"""
        if not mnemonic or not mnemonic.strip():
            return False

        words = mnemonic.strip().split()
        valid_lengths = [12, 15, 18, 21, 24]

        if len(words) not in valid_lengths:
            return False

        for word in words:
            if not word.isalpha():
                return False

        return True


class PasswordStrengthChecker:
    """Utility class for checking password strength."""

    @staticmethod
    def calculate_entropy(password: str) -> float:
        """Calculate estimated entropy bits for a password."""
        if not password:
            return 0.0

        charset_size = 0
        if any(c.islower() for c in password):
            charset_size += 26
        if any(c.isupper() for c in password):
            charset_size += 26
        if any(c.isdigit() for c in password):
            charset_size += 10
        if any(c in "!@#$%^&*()_+-=[]{}|;:,.<>?" for c in password):
            charset_size += 23

        if charset_size == 0:
            return 0.0

        import math
        return len(password) * math.log2(charset_size)

    @staticmethod
    def check_password_strength(password: str) -> dict:
        """
        Check password strength and return detailed analysis.

        Returns:
            Dictionary with strength analysis
        """
        if not password:
            return {
                'score': 0,
                'strength': 'Very Weak',
                'entropy': 0.0,
                'recommendations': ['Password cannot be empty']
            }

        issues = []
        score = 0

        # Length check
        if len(password) < 8:
            issues.append('Use at least 8 characters (12+ recommended)')
        elif len(password) < 12:
            issues.append('Consider using 12+ characters for better security')
            score += 1
        else:
            score += 2

        # Character variety checks
        if any(c.islower() for c in password):
            score += 1
        else:
            issues.append('Add lowercase letters')

        if any(c.isupper() for c in password):
            score += 1
        else:
            issues.append('Add uppercase letters')

        if any(c.isdigit() for c in password):
            score += 1
        else:
            issues.append('Add numbers')

        if any(c in "!@#$%^&*()_+-=[]{}|;:,.<>?" for c in password):
            score += 1
        else:
            issues.append('Add special characters (!@#$%^&*)')

        # Calculate entropy
        entropy = PasswordStrengthChecker.calculate_entropy(password)

        # Determine strength level
        if score >= 6 and entropy >= 70:
            strength = 'Very Strong'
        elif score >= 5 and entropy >= 60:
            strength = 'Strong'
        elif score >= 4 and entropy >= 50:
            strength = 'Moderate'
        elif score >= 2:
            strength = 'Weak'
        else:
            strength = 'Very Weak'

        return {
            'score': score,
            'strength': strength,
            'entropy': entropy,
            'recommendations': issues
        }

File: mobile/crypto/SecureEncryption.js
/**
 * Secure Mnemonic Encryption - JavaScript/React Native Version
 * Converted from Python secure_encryption.py
 * Uses same security standards: AES + HMAC with PBKDF2
 */

// IMPORTANT: Import polyfill first
import 'react-native-get-random-values';
import CryptoJS from 'crypto-js';
import * as Crypto from 'expo-crypto';

export class SecureMnemonicEncryption {
    constructor() {
        // Security constants optimized for mobile performance
        this.PBKDF2_ITERATIONS = 10000; // Reduced for mobile performance while still secure
        this.SALT_SIZE = 16; // 16 bytes = 128 bits
    }

    /**
     * Encrypt a mnemonic phrase with authenticated encryption
     * Equivalent to Python encrypt_mnemonic method
     *
     * @param {string} mnemonic - The mnemonic phrase to encrypt
     * @param {string} password - The encryption password
     * @returns {string} Base64 encoded string containing salt + encrypted data
     */
    encryptMnemonic(mnemonic, password) {
        if (!mnemonic || !mnemonic.trim()) {
            throw new Error("Mnemonic cannot be empty");
        }
        if (!password || password.length < 8) {
            throw new Error("Password must be at least 8 characters");
        }

        try {
            // Generate cryptographically secure random salt using simpler method
            const salt = CryptoJS.lib.WordArray.random(this.SALT_SIZE);

            // Derive key using PBKDF2 with high iteration count (same as Python)
            const key = CryptoJS.PBKDF2(password, salt, {
                keySize: 256/32, // 256 bits = 32 bytes
                iterations: this.PBKDF2_ITERATIONS,
                hasher: CryptoJS.algo.SHA256
            });

            // Encrypt with AES (CryptoJS uses authenticated encryption similar to Fernet)
            const encrypted = CryptoJS.AES.encrypt(mnemonic, key.toString());

            // Combine salt + encrypted data (same format as Python)
            const combined = salt.toString() + ':' + encrypted.toString();

            // Return base64 encoded result
            return CryptoJS.enc.Base64.stringify(CryptoJS.enc.Utf8.parse(combined));

        } catch (error) {
            throw new Error(`Encryption failed: ${error.message}`);
        }
    }

    /**
     * Decrypt an encrypted mnemonic phrase
     * Equivalent to Python decrypt_mnemonic method
     *
     * @param {string} encryptedData - Base64 encoded encrypted mnemonic
     * @param {string} password - The decryption password
     * @returns {string|null} Decrypted mnemonic phrase or null if decryption fails
     */
    decryptMnemonic(encryptedData, password) {
        try {
            if (!encryptedData || !password) {
                return null;
            }

            // Decode the base64 data
            const combined = CryptoJS.enc.Base64.parse(encryptedData).toString(CryptoJS.enc.Utf8);

            // Split salt and encrypted data
            const parts = combined.split(':');
            if (parts.length !== 2) {
                return null;
            }

            const saltHex = parts[0];
            const encryptedHex = parts[1];

            // Convert salt back to WordArray
            const salt = CryptoJS.enc.Hex.parse(saltHex);

            // Derive the same key using the extracted salt
            const key = CryptoJS.PBKDF2(password, salt, {
                keySize: 256/32,
                iterations: this.PBKDF2_ITERATIONS,
                hasher: CryptoJS.algo.SHA256
            });

            // Decrypt and verify integrity
            const decrypted = CryptoJS.AES.decrypt(encryptedHex, key.toString());
            const decryptedText = decrypted.toString(CryptoJS.enc.Utf8);

            if (!decryptedText) {
                return null; // Decryption failed or wrong password
            }

            return decryptedText;

        } catch (error) {
            // Don't reveal specific error details for security
            return null;
        }
    }

    /**
     * Basic validation of mnemonic format
     * Same logic as Python version
     *
     * @param {string} mnemonic - The mnemonic phrase to validate
     * @returns {boolean} True if format appears valid
     */
    verifyMnemonicFormat(mnemonic) {
        if (!mnemonic || !mnemonic.trim()) {
            return false;
        }

        const words = mnemonic.trim().split(/\s+/);

        // Standard mnemonic lengths: 12, 15, 18, 21, 24 words
        const validLengths = [12, 15, 18, 21, 24];

        if (!validLengths.includes(words.length)) {
            return false;
        }

        // Basic character validation (should only contain letters)
        for (const word of words) {
            if (!/^[a-zA-Z]+$/.test(word)) {
                return false;
            }
        }

        return true;
    }
}

export class PasswordStrengthChecker {
    /**
     * Calculate estimated entropy bits for a password
     * Same algorithm as Python version
     */
    static calculateEntropy(password) {
        if (!password) {
            return 0.0;
        }

        let charsetSize = 0;
        if (/[a-z]/.test(password)) charsetSize += 26;
        if (/[A-Z]/.test(password)) charsetSize += 26;
        if (/[0-9]/.test(password)) charsetSize += 10;
        if (/[!@#$%^&*()_+\-=\[\]{}|;:,.<>?]/.test(password)) charsetSize += 23;

        if (charsetSize === 0) {
            return 0.0;
        }

        return password.length * Math.log2(charsetSize);
    }

    /**
     * Check password strength and return detailed analysis
     * Same logic as Python version
     */
    static checkPasswordStrength(password) {
        if (!password) {
            return {
                score: 0,
                strength: 'Very Weak',
                entropy: 0.0,
                recommendations: ['Password cannot be empty']
            };
        }

        const issues = [];
        let score = 0;

        // Length check
        if (password.length < 8) {
            issues.push('Use at least 8 characters (12+ recommended)');
        } else if (password.length < 12) {
            issues.push('Consider using 12+ characters for better security');
            score += 1;
        } else {
            score += 2;
        }

        // Character variety checks
        if (/[a-z]/.test(password)) {
            score += 1;
        } else {
            issues.push('Add lowercase letters');
        }

        if (/[A-Z]/.test(password)) {
            score += 1;
        } else {
            issues.push('Add uppercase letters');
        }

        if (/[0-9]/.test(password)) {
            score += 1;
        } else {
            issues.push('Add numbers');
        }

        if (/[!@#$%^&*()_+\-=\[\]{}|;:,.<>?]/.test(password)) {
            score += 1;
        } else {
            issues.push('Add special characters (!@#$%^&*)');
        }

        // Calculate entropy
        const entropy = this.calculateEntropy(password);

        // Determine strength level
        let strength;
        if (score >= 6 && entropy >= 70) {
            strength = 'Very Strong';
        } else if (score >= 5 && entropy >= 60) {
            strength = 'Strong';
        } else if (score >= 4 && entropy >= 50) {
            strength = 'Moderate';
        } else if (score >= 2) {
            strength = 'Weak';
        } else {
            strength = 'Very Weak';
        }

        return {
            score,
            strength,
            entropy,
            recommendations: issues
        };
    }
}